File: CodeForces/Codeforces Round 1040 (Div 2)/D.cpp
#include <bits/stdc++.h>
#define int long long
using namespace std;
struct BIT {
    int n;
    vector<int> f;
    BIT(int _n=0) : n(_n) {
        f.assign(n+1, 0);
    }
    void init(int _n) {
        n=_n;
        f.assign(n+1, 0);
    }
    void update(int i) {
        for (; i<=n; i+=i&-i) {
            f[i]++;
        }
    }
    int query(int i) {
        int s=0;
        for (; i>0; i-=i&-i) {
            s+=f[i];
        }
        return s;
    }
};

int32_t main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int t;
    cin>>t;
    while(t--) {
        int n;
        cin>>n;
        vector<int> p(n+1);
        for (int i=1; i<=n; ++i) {
            cin>>p[i];
        }
        vector<int> invr(n+1), gl(n+1);
        BIT bit(n);
        int invp=0;
        bit.init( n);    
        for (int i=n; i>=1; i--) {
            invr[i]=bit.query(p[i]-1);
            invp+=invr[i];
            bit.update(p[i]);
        }
        bit.init( n);
        for (int i=1; i<=n; i++) {
            gl[i]=(i-1)-bit.query(p[i]-1);
            bit.update(p[i]);
        }
        long long sum_neg=0;
        for (int i=1; i<=n; ++i) {
            long long D=(n-i)-invr[i]-gl[i];
            if(D<0) {
                sum_neg+=D;
            }
        }
        long long ans=invp+sum_neg;
        cout<<ans<<'\n';
    }
    return 0;
}

File: CodeForces/Codeforces Round 1040 (Div 2)/D.py
import sys

def solve():
    n = int(sys.stdin.readline())
    p = [0] + list(map(int, sys.stdin.readline().split()))
    
    invr = [0] * (n + 1)
    gl = [0] * (n + 1)
    
    bit = Tree(2 * n + 1)
    
    invp = 0
    for i in range(n, 0, -1):
        invr[i] = bit.query(p[i] - 1)
        invp += invr[i]
        bit.upd(p[i])
        
    bit.init(2 * n + 1)
    
    for i in range(1, n + 1):
        gl[i] = bit.query(p[i] - 1)
        bit.upd(p[i])
    
    sum_neg = 0
    for i in range(1, n + 1):
        D = (n - p[i]) - invr[i] - gl[i]
        if D < 0:
            sum_neg += D
    
    ans = invp + sum_neg
    print(ans)

class Tree:
    def __init__(self, size):
        self.size = size
        self.tree = [0] * (size + 1)
    
    def init(self, size):
        self.size = size
        self.tree = [0] * (size + 1)

    def upd(self, index):
        while index <= self.size:
            self.tree[index] += 1
            index += index & (-index)
            
    def query(self, index):
        s = 0
        while index > 0:
            s += self.tree[index]
            index -= index & (-index)
        return s

t = int(sys.stdin.readline())
for _ in range(t):
    solve()